### PR TITLE
search for table

### PR DESCRIPTION
**Pull request by seryozhabaleyko**

### CODE REVIEW

File: src/components/TableWithSearch/TableWithSearch.vue [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+</template>
+
+<script lang="ts">
+import { defineComponent } from '@vue/composition-api';
+
+import BaseTable from '../Table';
+import Search from './components/Search';
+
+interface Props {
+  columnDefs: Array<{ [key: string]: unknown }>;
```
Comment (saranchonkau) [inline src/components/TableWithSearch/TableWithSearch.vue]: use existing type for ColumnDefinition

File: src/components/TableWithSearch/components/Search/Search.vue [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      type: Boolean,
+      default: false,
+    },
+  },
+  setup(props, context) {
+    const searchQuery = ref<string>(props.value);
+
+    const debounced = debounce(() => {
+      context.emit('input', searchQuery.value);
+    }, 500);
```
Comment (saranchonkau) [inline src/components/TableWithSearch/components/Search/Search.vue]: 300

File: src/typings/model.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  readonly price: number;
+
+  readonly image: Nullable<FileType>;
+  readonly video: Nullable<FileType>;
+
+  readonly platforms: Array<string>;
+
+  readonly supportedByVideoEditorApp: boolean;
+  readonly isBasePackage: boolean;
+} & SeoType;
```
Comment (saranchonkau) [inline src/typings/model.ts]: Remove this file

File: src/composables/useTableResource.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  context: SetupContext;
+}) {
+  const { context, fetchResource } = params;
+  const [searchQuery, setSearchQuery] = useSearch(context);
+  const isSearchLoading = ref<boolean>(false);
+
+  const [
+    doFetch,
+    { loading: isLoading, data: rowData, error: errorMessage },
+  ] = useResource<Array<ProductType>>({
```
Comment (saranchonkau) [inline src/composables/useTableResource.ts]: remove ProductType

File: src/components/TableWithSearch/hooks/useSearch.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      : context.root.$route.path;
+
+    if (context.root.$route.fullPath !== newUrl) {
+      context.root.$router.replace(newUrl);
+    }
+  }
+
+  function setSearchQuery(newQuery: string): void {
+    searchQuery.value = newQuery;
+    // updateSearchParam(newQuery);
```
Comment (saranchonkau) [inline src/components/TableWithSearch/hooks/useSearch.ts]: uncomment

File: src/components/TableWithSearch/components/Search/Search.vue [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    },
+    disabled: {
+      type: Boolean,
+      default: false,
+    },
+    readonly: {
+      type: Boolean,
+      default: false,
+    },
+    allowClear: {
```
Comment (saranchonkau) [inline src/components/TableWithSearch/components/Search/Search.vue]: remove prop

File: src/components/TableWithSearch/components/Search/Search.vue [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    },
+    placeholder: {
+      type: String,
+      default: 'Search...',
+    },
+    disabled: {
+      type: Boolean,
+      default: false,
+    },
+    readonly: {
```
Comment (saranchonkau) [inline src/components/TableWithSearch/components/Search/Search.vue]: remove

File: src/components/TableWithSearch/components/Search/Search.vue [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        aria-label="Clear search"
+        role="button"
+        @click="handleClear"
+      >
+        <SvgIcon
+          class="icon-search-off"
+          name="clear"
+          focusable="false"
+          aria-hidden="true"
+          role="img"
```
Comment (saranchonkau) [inline src/components/TableWithSearch/components/Search/Search.vue]: why we need it?

File: src/components/TableWithSearch/components/Search/Search.vue [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+      >
+        <Spinner
+          class="spinner"
+          size="24"
+          focusable="false"
+          aria-hidden="true"
+        />
+      </span>
+
+      <div
```
Comment (saranchonkau) [inline src/components/TableWithSearch/components/Search/Search.vue]: it must have `button` tag

File: src/components/Search/Search.vue [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        class="search-control"
+        type="text"
+        v-bind="$attrs"
+        :disabled="disabled"
+        :autocomplete="autocomplete"
+        :placeholder="placeholder"
+      />
+
+      <SvgIcon
+        v-if="!searchQuery && !loading"
```
Comment (saranchonkau) [inline src/components/Search/Search.vue]: do we really need this condition here? 
`!searchQuery`

File: src/components/DataTable/DataTable.hooks.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+interface TableState<T> {
+  isLoading: ComputedRef<boolean>;
+  rowData: Ref<Array<T>>;
+  errorMessage: Ref<Nullable<string>>;
+  searchQuery: Ref<string>;
+  setSearchQuery: (newSearchQuery: string) => void;
+  handleChange: (event: TableChangeEvent) => void;
+  fetchEntityList: () => void;
```
Comment (saranchonkau) [inline src/components/DataTable/DataTable.hooks.ts]: wrong type
should be `(params: TableDataRequestParams ) => void;`

File: src/components/Search/Search.vue [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    },
+    loading: {
+      type: Boolean,
+      default: false,
+    },
+  },
+  setup(props, context) {
+    const searchQuery = ref<string>(props.value);
+
+    const debounced = debounce(() => {
```
Comment (saranchonkau) [inline src/components/Search/Search.vue]: rename to `emitDebouncedChangeEvent`

File: src/components/Search/Search.vue (hunk 10 above)
Comment (seryozhabaleyko) [reply src/components/Search/Search.vue]: Yes, this is a search icon and should not be shown when text is entered.

**Pull request merged**